nmtwizard/preprocess.py

```diff
@@ -5,13 +5,61 @@
 from nmtwizard.logger import get_logger
 from nmtwizard.sampler import sample
 from nmtwizard import prepoperator
+from nmtwizard import tokenizer
 
 logger = get_logger(__name__)
 
-def generate_preprocessed_data(config, corpus_dir, data_dir):
+def _generate_models(config, corpus_dir, data_dir, option):
 
-    # TODO : annotations
-    # TODO : file-specific rules/extra
+    opt_multi = config.get('tokenization', {}).get('multi', {}).get(option)
+    opt_source = config.get('tokenization', {}).get('source', {}).get(option)
+    opt_target = config.get('tokenization', {}).get('target', {}).get(option)
+
+    if not opt_multi and not (opt_source and opt_target):
+        logger.warning('No \'' + option + '\' option specified, exit without preprocessing.')
+        return
+
+    if (opt_multi and opt_source) or \
+       (opt_multi and opt_target) :
+        raise RuntimeError('Cannot specify \'' + option + '\' for both \'multi\' and either \'source\' or \'target\'.')
+
+    # Generate preprocessed sentences and feed them to subword learners or to vocabularies.
+    generate_preprocessed_data(config, corpus_dir, data_dir, option)
+
+
+def generate_vocabularies(config, corpus_dir, data_dir):
+
+    # TODO V2 : change this when tokenization will be part of preprocessing pipeline.
+    if 'tokenization' not in config:
+        raise RuntimeError('No \'tokenization\' field in configuration, \
+                            cannot build the vocabularies.)')
+
+    if ('source' not in config['tokenization'] or \
+        'target' not in config['tokenization']) and \
+        'multi' not in config['tokenization'] :
+        raise RuntimeError('\'tokenization\' field should contain \
```

Comment:
If no "source" or "target" are present, could we simply assume that the full block is "multi", i.e we could have:

```text
{"source": {source options...}, "target": {target options}}
```
or
```text
{multi options...}
```
?

Some parts of the code make it look like you can have "multi", "source", and "target" at the same time.

Reply:
You are right as usual :)
In fact, "multi" section was reserved in my original thinking for options that imperatively require two sides, like alignment, entity checks, etc...

Following this logic, options for building joint subword models or joint vocabularies should be placed in “multi” section, since they require both sides to be executed.

Tokenization options are, on the hand, intrinsically “monolingual”, applied to each side separately. Even if they are the same for both sides, they do not really require both sides. Thus there should not be “multi” section for tokenization itself.

To avoid the confusion, I separate “build_vocabulary” et “build_subword” options (these can be present either in both “source” and “target” or in “multi”.
The results (subword model or vocabulary),  on the other hand, are always put in “source” and “target” separately. So are all the other tokenization options.

Please let me know if this looks ok for you.

Reply:
Sorry I missed your reply. Thanks, it makes sense.

Reply:
Sorry I missed your reply. Thanks, it makes sense.